source/validate_datarules.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2015-2016 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Performs validation on instructions that appear inside of a SPIR-V block.
+
+#include "validate.h"
+
+#include <cassert>
+#include <sstream>
+#include <string>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "operand.h"
+#include "val/ValidationState.h"
+
+using libspirv::CapabilitySet;
+using libspirv::DiagnosticStream;
+using libspirv::ValidationState_t;
+
+namespace {
+
+// vector types can only have 2 or 3 or 4 components.
+// if Vector16 capability is added, 8 and 16 components are also allowed.
+spv_result_t ValidateNumVecComponents(ValidationState_t& _,
+                                      const spv_parsed_instruction_t* inst) {
+  if (inst->opcode == SpvOpTypeVector) {
+    const std::string opcode_string =
```

Comment:
It's wasteful to construct this string in the valid cases.  I'd just defer the call until the error message generation. Even if that means repeating your self. Repeating yourself is not the worst thing in the world.